File: src/lib/scheduler.py
"""Scheduled loop support for Zephyr Desktop.

Provides LoopScheduler, which manages timer-based periodic execution
of Ralph loops using simplified cron-like expressions.  Runs entirely
in-process via threading.Timer (no external cron daemon required).

Supported schedule expressions:
    "*/N minutes"  — every N minutes
    "every N hours" — every N hours
    "hourly"       — alias for "every 1 hours"
    "daily HH:MM"  — once per day at the given UTC time
"""

import logging
import re
import threading
from datetime import datetime, timezone
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from src.lib.loop_runner import LoopRunner

from src.lib.loop_runner import LoopMode

logger = logging.getLogger("zephyr.scheduler")


# ---------------------------------------------------------------------------
# Schedule expression parser
# ---------------------------------------------------------------------------

# */5 minutes, */30 minutes, etc.
_RE_EVERY_N_MINUTES = re.compile(r"^\*/(\d+)\s+minutes?$", re.IGNORECASE)

# every 2 hours, every 1 hours, etc.
_RE_EVERY_N_HOURS = re.compile(r"^every\s+(\d+)\s+hours?$", re.IGNORECASE)

# daily 14:30, daily 08:00, etc.
_RE_DAILY = re.compile(r"^daily\s+(\d{1,2}):(\d{2})$", re.IGNORECASE)

# hourly (alias)
_RE_HOURLY = re.compile(r"^hourly$", re.IGNORECASE)


def parse_schedule(expr: str) -> float:
    """Parse a simplified cron expression and return the interval in seconds.

    For ``"daily HH:MM"`` expressions the returned value is the number of
    seconds from *now* until the next occurrence of that wall-clock time
    (in UTC).  For all other expressions the value is a fixed repeat
    interval.

    Raises:
        ValueError: If the expression cannot be parsed.
    """
    expr = expr.strip()

    m = _RE_EVERY_N_MINUTES.match(expr)
    if m:
        minutes = int(m.group(1))
        if minutes <= 0:
            raise ValueError(f"Minutes must be > 0, got {minutes}")
        return minutes * 60.0

    m = _RE_HOURLY.match(expr)
    if m:
        return 3600.0

    m = _RE_EVERY_N_HOURS.match(expr)
    if m:
        hours = int(m.group(1))
        if hours <= 0:
            raise ValueError(f"Hours must be > 0, got {hours}")
        return hours * 3600.0

    m = _RE_DAILY.match(expr)
    if m:
        hour = int(m.group(1))
        minute = int(m.group(2))
        if not (0 <= hour <= 23 and 0 <= minute <= 59):
            raise ValueError(
                f"Invalid time in daily expression: {hour:02d}:{minute:02d}"
            )
        return _seconds_until_daily(hour, minute)

    raise ValueError(f"Unrecognised schedule expression: {expr!r}")


def _seconds_until_daily(hour: int, minute: int) -> float:
    """Return seconds from now until the next occurrence of HH:MM UTC."""
    now = datetime.now(timezone.utc)
    target = now.replace(hour=hour, minute=minute, second=0, microsecond=0)
    delta = (target - now).total_seconds()
    if delta <= 0:
        # Already past today's occurrence — schedule for tomorrow
        delta += 86400.0
    return delta


# ---------------------------------------------------------------------------
# LoopScheduler
# ---------------------------------------------------------------------------


class LoopScheduler:
    """Timer-based scheduler for periodic loop execution.

    Uses ``threading.Timer`` with a re-scheduling pattern so that each
    firing sets up the next one, avoiding drift from long-running loop
    starts.

    Args:
        loop_runner: The LoopRunner used to start loops.
    """

    def __init__(self, loop_runner: "LoopRunner") -> None:
        self._loop_runner = loop_runner
        self._schedules: dict[str, str] = {}  # project_id -> cron_expr
        self._timers: dict[str, threading.Timer] = {}
        self._lock = threading.Lock()

    # -- Public API ----------------------------------------------------------

    def schedule_loop(self, project_id: str, cron_expr: str) -> None:
        """Schedule periodic loop execution for a project.

        Parses *cron_expr* to determine the interval, then starts a
        ``threading.Timer`` that fires ``loop_runner.start_loop`` at
        the appropriate time.  Subsequent firings are re-scheduled
        automatically.

        Raises:
            ValueError: If *cron_expr* cannot be parsed or a schedule
                already exists for this project.
        """
        interval = parse_schedule(cron_expr)  # validates expression

        with self._lock:
            if project_id in self._schedules:
                raise ValueError(
                    f"Schedule already exists for project '{project_id}'. "
                    "Cancel it first."
                )
            self._schedules[project_id] = cron_expr

        self._start_timer(project_id, cron_expr, interval)
        logger.info(
            "Scheduled loop for project %s: %s (first fire in %.0fs)",
            project_id,
            cron_expr,
            interval,
        )

    def cancel_schedule(self, project_id: str) -> None:
        """Cancel a previously scheduled loop.

        Raises:
            KeyError: If no schedule exists for the project.
        """
        with self._lock:
            if project_id not in self._schedules:
                raise KeyError(f"No schedule found for project '{project_id}'")
            del self._schedules[project_id]
            timer = self._timers.pop(project_id, None)

        if timer is not None:
            timer.cancel()
            logger.info("Cancelled schedule for project %s", project_id)

    def list_schedules(self) -> dict[str, str]:
        """Return a copy of ``{project_id: cron_expr}`` for all active schedules."""
        with self._lock:
            return dict(self._schedules)

    # -- Internal helpers ----------------------------------------------------

    def _start_timer(self, project_id: str, cron_expr: str, interval: float) -> None:
        """Create and start a Timer that will fire ``_on_timer``."""
        timer = threading.Timer(interval, self._on_timer, args=(project_id, cron_expr))
        timer.daemon = True
        timer.name = f"zephyr-sched-{project_id[:12]}"

        with self._lock:
            # Replace any stale timer reference
            old = self._timers.pop(project_id, None)
            if old is not None:
                old.cancel()
            self._timers[project_id] = timer

        timer.start()

    def _on_timer(self, project_id: str, cron_expr: str) -> None:
        """Callback executed when a scheduled timer fires.

        Starts the loop via LoopRunner and, if the schedule is still
        active, re-schedules the next occurrence.
        """
        with self._lock:
            if project_id not in self._schedules:
                # Schedule was cancelled between fire and callback
                return

        # Attempt to start the loop
        try:
            self._loop_runner.start_loop(
                project_id=project_id,
                mode=LoopMode.SCHEDULED,
            )
            logger.info("Scheduled fire: started loop for project %s", project_id)
        except Exception as exc:
            logger.error(
                "Scheduled fire failed for project %s: %s",
                project_id,
                exc,
            )

        # Re-schedule if still active
        with self._lock:
            if project_id not in self._schedules:
                return

        try:
            next_interval = parse_schedule(cron_expr)
        except ValueError:
            logger.error(
                "Failed to re-parse schedule '%s' for project %s",
                cron_expr,
                project_id,
            )
            return

        self._start_timer(project_id, cron_expr, next_interval)

File: src/services/scheduler.ts
/**
 * Loop scheduling service for Zephyr Desktop.
 *
 * LoopScheduler manages recurring loop executions based on schedule expressions.
 * Supports minute-based intervals, hourly intervals, and daily fixed-time schedules.
 */

import type { LoopRunner } from './loop-runner';
import type { LoopStartOpts } from '../shared/loop-types';
import { LoopMode } from '../shared/loop-types';

/**
 * Parsed schedule configuration.
 */
export interface ScheduleConfig {
  /** Interval in milliseconds, or null for daily fixed-time schedules */
  intervalMs: number | null;

  /** For daily schedules: hour (0-23) */
  dailyHour?: number;

  /** For daily schedules: minute (0-59) */
  dailyMinute?: number;

  /** Original expression string */
  expression: string;
}

/**
 * Active scheduled loop entry.
 */
export interface ScheduledLoop {
  /** Project ID being scheduled */
  projectId: string;

  /** Schedule configuration */
  schedule: ScheduleConfig;

  /** Loop start options used for each trigger */
  loopOpts: Omit<LoopStartOpts, 'mode'>;

  /** Timer ID (NodeJS.Timeout or NodeJS.Timer) */
  timerId: NodeJS.Timeout | null;

  /** ISO 8601 timestamp of next scheduled run */
  nextRun: string | null;
}

/**
 * Manages scheduled loop executions.
 *
 * Responsibilities:
 * - Parse schedule expressions into timer configurations
 * - Schedule loops to run at intervals or fixed times
 * - Trigger LoopRunner.startLoop() with mode=SINGLE at scheduled times
 * - Manage active timers and cancellation
 */
export class LoopScheduler {
  private loopRunner: LoopRunner;
  private schedules: Map<string, ScheduledLoop> = new Map();

  constructor(loopRunner: LoopRunner) {
    this.loopRunner = loopRunner;
  }

  /**
   * Parse a schedule expression into a ScheduleConfig.
   *
   * Supported formats:
   * - Minute intervals: every N minutes
   * - Hour intervals: every N hours
   * - Daily schedules: daily at fixed time HH:MM
   *
   * @param expr - Schedule expression string
   * @returns Parsed schedule configuration
   * @throws Error if expression format is invalid
   */
  parseSchedule(expr: string): ScheduleConfig {
    const trimmed = expr.trim().toLowerCase();

    // Pattern 1: "*/N minutes"
    const minuteMatch = trimmed.match(/^\*\/(\d+)\s+minutes?$/);
    if (minuteMatch) {
      const minutes = parseInt(minuteMatch[1], 10);
      if (minutes < 1) {
        throw new Error('Minute interval must be at least 1');
      }
      return {
        intervalMs: minutes * 60 * 1000,
        expression: expr,
      };
    }

    // Pattern 2: "every N hours"
    const hourMatch = trimmed.match(/^every\s+(\d+)\s+hours?$/);
    if (hourMatch) {
      const hours = parseInt(hourMatch[1], 10);
      if (hours < 1) {
        throw new Error('Hour interval must be at least 1');
      }
      return {
        intervalMs: hours * 60 * 60 * 1000,
        expression: expr,
      };
    }

    // Pattern 3: "daily HH:MM"
    const dailyMatch = trimmed.match(/^daily\s+(-?\d{1,2}):(-?\d{1,2})$/);
    if (dailyMatch) {
      const hour = parseInt(dailyMatch[1], 10);
      const minute = parseInt(dailyMatch[2], 10);

      if (hour < 0 || hour > 23) {
        throw new Error('Hour must be between 0 and 23');
      }
      if (minute < 0 || minute > 59) {
        throw new Error('Minute must be between 0 and 59');
      }

      return {
        intervalMs: null,
        dailyHour: hour,
        dailyMinute: minute,
        expression: expr,
      };
    }

    throw new Error(
      `Invalid schedule expression: "${expr}". ` +
        'Expected formats: "*/N minutes", "every N hours", or "daily HH:MM"'
    );
  }

  /**
   * Schedule a loop to run at regular intervals or fixed times.
   *
   * @param projectId - UUID of the project to schedule
   * @param schedule - Schedule expression or parsed ScheduleConfig
   * @param loopOpts - Loop start options (excluding mode, which is forced to SINGLE)
   * @throws Error if schedule expression is invalid or project already scheduled
   */
  scheduleLoop(
    projectId: string,
    schedule: string | ScheduleConfig,
    loopOpts: Omit<LoopStartOpts, 'mode'>
  ): void {
    if (this.schedules.has(projectId)) {
      throw new Error(`Project ${projectId} is already scheduled`);
    }

    const config =
      typeof schedule === 'string' ? this.parseSchedule(schedule) : schedule;

    const scheduled: ScheduledLoop = {
      projectId,
      schedule: config,
      loopOpts,
      timerId: null,
      nextRun: null,
    };

    if (config.intervalMs !== null) {
      // Interval-based schedule (minutes or hours)
      scheduled.nextRun = new Date(Date.now() + config.intervalMs).toISOString();
      scheduled.timerId = setInterval(() => {
        this.triggerLoop(projectId);
        const sched = this.schedules.get(projectId);
        if (sched && sched.schedule.intervalMs !== null) {
          sched.nextRun = new Date(
            Date.now() + sched.schedule.intervalMs
          ).toISOString();
        }
      }, config.intervalMs);
    } else {
      // Daily fixed-time schedule
      scheduled.nextRun = this.calculateNextDailyRun(
        config.dailyHour!,
        config.dailyMinute!
      );
      this.scheduleDailyTimer(projectId);
    }

    this.schedules.set(projectId, scheduled);
  }

  /**
   * Cancel a scheduled loop.
   *
   * @param projectId - UUID of the project to cancel
   * @returns True if schedule was cancelled, false if not scheduled
   */
  cancelSchedule(projectId: string): boolean {
    const scheduled = this.schedules.get(projectId);
    if (!scheduled) {
      return false;
    }

    if (scheduled.timerId) {
      clearInterval(scheduled.timerId);
      clearTimeout(scheduled.timerId);
    }

    this.schedules.delete(projectId);
    return true;
  }

  /**
   * List all currently scheduled loops.
   *
   * @returns Array of ScheduledLoop entries (without sensitive data)
   */
  listScheduled(): ScheduledLoop[] {
    return Array.from(this.schedules.values());
  }

  /**
   * Check if a project is currently scheduled.
   *
   * @param projectId - UUID of the project
   * @returns True if scheduled, false otherwise
   */
  isScheduled(projectId: string): boolean {
    return this.schedules.has(projectId);
  }

  /**
   * Cancel all scheduled loops (used during shutdown).
   */
  cancelAll(): void {
    for (const projectId of this.schedules.keys()) {
      this.cancelSchedule(projectId);
    }
  }

  /**
   * Trigger a loop execution immediately for a scheduled project.
   *
   * @param projectId - UUID of the project
   */
  private async triggerLoop(projectId: string): Promise<void> {
    const scheduled = this.schedules.get(projectId);
    if (!scheduled) {
      return;
    }

    try {
      // Start loop with mode=SINGLE (scheduler always triggers single-shot loops)
      await this.loopRunner.startLoop({
        ...scheduled.loopOpts,
        mode: LoopMode.SINGLE,
      });
    } catch (err) {
      // Log error but don't cancel schedule (user may fix issue and next run succeeds)
      console.error(
        `[Scheduler] Failed to trigger loop for project ${projectId}:`,
        err
      );
    }
  }

  /**
   * Calculate the next daily run time.
   *
   * @param hour - Target hour (0-23)
   * @param minute - Target minute (0-59)
   * @returns ISO 8601 timestamp of next run
   */
  private calculateNextDailyRun(hour: number, minute: number): string {
    const now = new Date();
    const next = new Date();
    next.setUTCHours(hour, minute, 0, 0);

    // If target time is in the past today, schedule for tomorrow
    if (next <= now) {
      next.setUTCDate(next.getUTCDate() + 1);
    }

    return next.toISOString();
  }

  /**
   * Schedule or reschedule a daily timer.
   *
   * @param projectId - UUID of the project
   */
  private scheduleDailyTimer(projectId: string): void {
    const scheduled = this.schedules.get(projectId);
    if (!scheduled || scheduled.schedule.intervalMs !== null) {
      return;
    }

    const { dailyHour, dailyMinute } = scheduled.schedule;
    const nextRun = new Date(
      this.calculateNextDailyRun(dailyHour!, dailyMinute!)
    );
    const msUntilRun = nextRun.getTime() - Date.now();

    if (scheduled.timerId) {
      clearTimeout(scheduled.timerId);
    }

    scheduled.timerId = setTimeout(() => {
      this.triggerLoop(projectId);

      // Reschedule for next day
      const sched = this.schedules.get(projectId);
      if (sched && sched.schedule.intervalMs === null) {
        sched.nextRun = this.calculateNextDailyRun(
          sched.schedule.dailyHour!,
          sched.schedule.dailyMinute!
        );
        this.scheduleDailyTimer(projectId);
      }
    }, msUntilRun);

    scheduled.nextRun = nextRun.toISOString();
  }
}
